src/pgduckdb_planner.cpp

```diff
@@ -6,34 +6,107 @@
 #include "pgduckdb/scan/postgres_scan.hpp"
 #include "pgduckdb/pgduckdb_types.hpp"
 #include "pgduckdb/pgduckdb_planner.hpp"
+#include "pgduckdb/pgduckdb_table_am.hpp"
 
 extern "C" {
 #include "postgres.h"
+
 #include "access/xact.h"
+#include "access/table.h"
 #include "catalog/pg_type.h"
 #include "nodes/makefuncs.h"
 #include "nodes/nodes.h"
+#include "nodes/nodeFuncs.h"
 #include "nodes/params.h"
 #include "optimizer/optimizer.h"
 #include "optimizer/planner.h"
 #include "optimizer/planmain.h"
+#include "parser/parse_coerce.h"
+#include "rewrite/rewriteHandler.h"
 #include "tcop/pquery.h"
+#include "utils/typcache.h"
+#include "utils/rel.h"
 #include "utils/syscache.h"
 #include "utils/guc.h"
 
 #include "pgduckdb/pgduckdb_ruleutils.h"
 }
 
+#include "pgduckdb/pg/types.hpp"
 #include "pgduckdb/pgduckdb_duckdb.hpp"
 #include "pgduckdb/pgduckdb_node.hpp"
 #include "pgduckdb/vendor/pg_list.hpp"
 #include "pgduckdb/utility/cpp_wrapper.hpp"
 #include "pgduckdb/pgduckdb_types.hpp"
 
+bool
+IsAllowedPostgresInsert(Query *query, bool throw_error) {
+	int elevel = throw_error ? ERROR : DEBUG4;
+	if (query->commandType != CMD_INSERT) {
+		elog(elevel, "DuckDB only supports INSERT/SELECT on Postgres tables");
+		return false;
+	}
+
+	Assert(list_length(query->rtable) >= query->resultRelation);
+	RangeTblEntry *target_rel = (RangeTblEntry *)list_nth(query->rtable, query->resultRelation - 1);
+	if (pgduckdb::IsDuckdbTable(target_rel->relid)) {
+		return false;
+	}
+
+	/* Checking supported INSERT types */
+	RangeTblEntry *select_rte = NULL;
+	foreach_node(RangeTblEntry, rte, query->rtable) {
+		if (rte->rtekind == RTE_SUBQUERY) {
+			select_rte = rte;
+		}
+	}
+
+	if (!select_rte) {
+		elog(elevel, "DuckDB does not support INSERT without a subquery");
```

Comment:
There are no tests for this error case afaict

Reply:
Added